File: comic_editor.cs
// comic_editor.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

class CharacterData {
    public string Name { get; set; }
    public Dictionary<string, string> Emojis { get; set; }
    public string Desc { get; set; }
}

class SceneItem {
    public string Character { get; set; }
    public string Name { get; set; }
    public string Emotion { get; set; }
    public string Dialogue { get; set; }
    public string Ascii { get; set; }
}

class ComicEditor {
    private Dictionary<string, CharacterData> characters = new Dictionary<string, CharacterData>();
    private List<SceneItem> scene = new List<SceneItem>();

    public ComicEditor() {
        InitCharacters();
    }

    private void InitCharacters() {
        characters["hero"] = new CharacterData {
            Name = "Hero",
            Emojis = new Dictionary<string, string> {
                ["default"] = "(^_^)\n /|\\\n / \\",
                ["радость"] = "(^_^)\n /|\\\n / \\",
                ["грусть"] = "(T_T)\n /|\\\n / \\",
                ["злость"] = "(>_<)\n /|\\\n / \\",
                ["удивление"] = "(O_O)\n /|\\\n / \\",
                ["страх"] = "(>_<)\n /|\\\n / \\"
            },
            Desc = "Отважный герой"
        };
        characters["villain"] = new CharacterData {
            Name = "Villain",
            Emojis = new Dictionary<string, string> {
                ["default"] = "(-_-)\n /|\\\n / \\",
                ["радость"] = "(^_^)\n /|\\\n / \\",
                ["грусть"] = "(T_T)\n /|\\\n / \\",
                ["злость"] = "(>_<)\n /|\\\n / \\",
                ["удивление"] = "(O_O)\n /|\\\n / \\",
                ["страх"] = "(>_<)\n /|\\\n / \\"
            },
            Desc = "Коварный злодей"
        };
        characters["robot"] = new CharacterData {
            Name = "Robot",
            Emojis = new Dictionary<string, string> {
                ["default"] = "[0_0]\n /|\\\n / \\",
                ["радость"] = "[^_^]\n /|\\\n / \\",
                ["грусть"] = "[T_T]\n /|\\\n / \\",
                ["злость"] = "[>_<]\n /|\\\n / \\",
                ["удивление"] = "[O_O]\n /|\\\n / \\",
                ["страх"] = "[>_<]\n /|\\\n / \\"
            },
            Desc = "Механический робот"
        };
        characters["monster"] = new CharacterData {
            Name = "Monster",
            Emojis = new Dictionary<string, string> {
                ["default"] = "({0_0})\n /|\\\n / \\",
                ["радость"] = "({^_^})\n /|\\\n / \\",
                ["грусть"] = "({T_T})\n /|\\\n / \\",
                ["злость"] = "({>_<})\n /|\\\n / \\",
                ["удивление"] = "({O_O})\n /|\\\n / \\",
                ["страх"] = "({>_<})\n /|\\\n / \\"
            },
            Desc = "Страшный монстр"
        };
        characters["princess"] = new CharacterData {
            Name = "Princess",
            Emojis = new Dictionary<string, string> {
                ["default"] = "(✿◠‿◠)\n /|\\\n / \\",
                ["радость"] = "(✿◠‿◠)\n /|\\\n / \\",
                ["грусть"] = "(✿T_T)\n /|\\\n / \\",
                ["злость"] = "(✿>_<)\n /|\\\n / \\",
                ["удивление"] = "(✿O_O)\n /|\\\n / \\",
                ["страх"] = "(✿>_<)\n /|\\\n / \\"
            },
            Desc = "Прекрасная принцесса"
        };
        characters["knight"] = new CharacterData {
            Name = "Knight",
            Emojis = new Dictionary<string, string> {
                ["default"] = "[⚔️] \n /|\\\n / \\",
                ["радость"] = "[⚔️^_^]\n /|\\\n / \\",
                ["грусть"] = "[⚔️T_T]\n /|\\\n / \\",
                ["злость"] = "[⚔️>_<]\n /|\\\n / \\",
                ["удивление"] = "[⚔️O_O]\n /|\\\n / \\",
                ["страх"] = "[⚔️>_<]\n /|\\\n / \\"
            },
            Desc = "Доблестный рыцарь"
        };
    }

    public void ListCharacters() {
        Console.WriteLine("\u001B[36mДоступные персонажи:\u001B[0m");
        int i = 1;
        foreach (var entry in characters) {
            Console.WriteLine($"  {i}. {entry.Value.Name} — {entry.Value.Desc}");
            i++;
        }
    }

    public string GetCharacter(int choice) {
        var keys = characters.Keys.ToList();
        if (choice >= 1 && choice <= keys.Count) {
            return keys[choice - 1];
        }
        return null;
    }

    public void CreateScene() {
        Console.WriteLine("\u001B[36m🎭 Создание сцены комикса\u001B[0m");
        ListCharacters();

        while (true) {
            Console.Write("Выберите персонажа (0 — завершить): ");
            string input = Console.ReadLine().Trim();
            if (!int.TryParse(input, out int choice)) {
                Console.WriteLine("\u001B[31m❌ Неверный выбор.\u001B[0m");
                continue;
            }
            if (choice == 0) break;

            string charKey = GetCharacter(choice);
            if (charKey == null) {
                Console.WriteLine("\u001B[31m❌ Неверный выбор.\u001B[0m");
                continue;
            }

            var charData = characters[charKey];
            Console.WriteLine($"Доступные эмоции: {string.Join(", ", charData.Emojis.Keys)}");
            Console.Write("Выберите эмоцию: ");
            string emotion = Console.ReadLine().Trim().ToLower();
            if (!charData.Emojis.ContainsKey(emotion)) {
                emotion = "default";
            }

            Console.Write("Введите реплику персонажа: ");
            string dialogue = Console.ReadLine().Trim();

            scene.Add(new SceneItem {
                Character = charKey,
                Name = charData.Name,
                Emotion = emotion,
                Dialogue = dialogue,
                Ascii = charData.Emojis[emotion]
            });
            Console.WriteLine($"\u001B[32m✅ {charData.Name} добавлен в сцену!\u001B[0m");
        }
    }

    public void RenderScene() {
        if (scene.Count == 0) {
            Console.WriteLine("\u001B[33mСцена пуста. Добавьте персонажей.\u001B[0m");
            return;
        }

        Console.WriteLine("\n+" + new string('-', 48) + "+");
        foreach (var item in scene) {
            foreach (var line in item.Ascii.Split('\n')) {
                Console.WriteLine($"| {line,-46} |");
            }
            if (!string.IsNullOrEmpty(item.Dialogue)) {
                Console.WriteLine($"| \u001B[33m💬 {item.Dialogue,-44}\u001B[0m |");
            }
            Console.WriteLine("|" + new string('-', 48) + "|");
        }
        Console.WriteLine("+" + new string('-', 48) + "+");
    }

    public void SaveHTML(string filename = "comic.html") {
        if (scene.Count == 0) {
            Console.WriteLine("\u001B[33mНет сцены для сохранения.\u001B[0m");
            return;
        }

        var html = new StringBuilder();
        html.AppendLine("<!DOCTYPE html>");
        html.AppendLine("<html><head><meta charset=\"UTF-8\"><title>Мой комикс</title>");
        html.AppendLine("<style>");
        html.AppendLine("body { font-family: monospace; background: #f0f0f0; padding: 20px; }");
        html.AppendLine(".panel { background: white; border: 2px solid #333; border-radius: 10px; padding: 20px; margin: 20px auto; max-width: 600px; box-shadow: 0 4px 8px rgba(0,0,0,0.1); }");
        html.AppendLine(".character { white-space: pre; font-size: 14px; line-height: 1.2; text-align: center; }");
        html.AppendLine(".dialogue { font-size: 18px; color: #2c3e50; text-align: center; margin: 10px 0; padding: 10px; background: #ecf0f1; border-radius: 10px; }");
        html.AppendLine(".name { font-weight: bold; color: #2980b9; text-align: center; }");
        html.AppendLine("</style></head><body><h1 style=\"text-align:center;\">🎭 Мой комикс</h1>");

        foreach (var item in scene) {
            html.AppendLine("<div class=\"panel\">");
            html.AppendLine($"    <div class=\"name\">{item.Name} ({item.Emotion})</div>");
            html.AppendLine($"    <div class=\"character\">{item.Ascii}</div>");
            html.AppendLine($"    <div class=\"dialogue\">💬 {item.Dialogue}</div>");
            html.AppendLine("</div>");
        }
        html.AppendLine("</body></html>");
        File.WriteAllText(filename, html.ToString());
        Console.WriteLine($"\u001B[32m💾 Комикс сохранён в {filename}\u001B[0m");
    }

    public void SaveJSON(string filename = "comic_project.json") {
        var sb = new StringBuilder();
        sb.AppendLine("{");
        sb.AppendLine($"  \"timestamp\": \"{DateTime.Now:yyyy-MM-ddTHH:mm:ss.fffZ}\",");
        sb.AppendLine("  \"scene\": [");
        for (int i = 0; i < scene.Count; i++) {
            var item = scene[i];
            sb.AppendLine("    {");
            sb.AppendLine($"      \"character\": \"{item.Character}\",");
            sb.AppendLine($"      \"name\": \"{item.Name}\",");
            sb.AppendLine($"      \"emotion\": \"{item.Emotion}\",");
            sb.AppendLine($"      \"dialogue\": \"{item.Dialogue}\",");
            sb.AppendLine($"      \"ascii\": \"{item.Ascii.Replace("\n", "\\n")}\"");
            sb.Append("    }");
            if (i < scene.Count - 1) sb.Append(",");
            sb.AppendLine();
        }
        sb.AppendLine("  ]");
        sb.AppendLine("}");
        File.WriteAllText(filename, sb.ToString());
        Console.WriteLine($"\u001B[32m💾 Проект сохранён в {filename}\u001B[0m");
    }

    public void LoadJSON(string filename = "comic_project.json") {
        if (!File.Exists(filename)) {
            Console.WriteLine($"\u001B[31m❌ Файл {filename} не найден.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[32m✅ Проект загружен из {filename}\u001B[0m");
        Console.WriteLine("\u001B[33m⚠️ Для полной загрузки используйте JSON-библиотеку.\u001B[0m");
    }

    public static void Main() {
        var editor = new ComicEditor();
        while (true) {
            Console.WriteLine("\n\u001B[36m🎭 Comic Editor Pro (C#)\u001B[0m");
            Console.WriteLine("1. Создать новую сцену");
            Console.WriteLine("2. Показать сцену");
            Console.WriteLine("3. Сохранить как HTML");
            Console.WriteLine("4. Сохранить проект (JSON)");
            Console.WriteLine("5. Загрузить проект (JSON)");
            Console.WriteLine("6. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine().Trim();

            switch (choice) {
                case "1": editor.CreateScene(); break;
                case "2": editor.RenderScene(); break;
                case "3": editor.SaveHTML(); break;
                case "4": editor.SaveJSON(); break;
                case "5": editor.LoadJSON(); break;
                case "6": Console.WriteLine("До свидания!"); return;
                default: Console.WriteLine("\u001B[31m❌ Неверный выбор.\u001B[0m"); break;
            }
        }
    }
}

File: comic_editor.java
// comic_editor.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.*;

class CharacterData {
    String name;
    Map<String, String> emojis;
    String desc;

    CharacterData(String name, Map<String, String> emojis, String desc) {
        this.name = name;
        this.emojis = emojis;
        this.desc = desc;
    }
}

class SceneItem {
    String character;
    String name;
    String emotion;
    String dialogue;
    String ascii;

    SceneItem(String character, String name, String emotion, String dialogue, String ascii) {
        this.character = character;
        this.name = name;
        this.emotion = emotion;
        this.dialogue = dialogue;
        this.ascii = ascii;
    }
}

public class comic_editor {
    private static Map<String, CharacterData> characters = new LinkedHashMap<>();
    private static List<SceneItem> scene = new ArrayList<>();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        initCharacters();
        while (true) {
            System.out.println("\n\u001B[36m🎭 Comic Editor Pro (Java)\u001B[0m");
            System.out.println("1. Создать новую сцену");
            System.out.println("2. Показать сцену");
            System.out.println("3. Сохранить как HTML");
            System.out.println("4. Сохранить проект (JSON)");
            System.out.println("5. Загрузить проект (JSON)");
            System.out.println("6. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine().trim();

            switch (choice) {
                case "1": createScene(); break;
                case "2": renderScene(); break;
                case "3": saveHTML(); break;
                case "4": saveJSON(); break;
                case "5": loadJSON(); break;
                case "6": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31m❌ Неверный выбор.\u001B[0m");
            }
        }
    }

    private static void initCharacters() {
        Map<String, String> heroEmojis = new LinkedHashMap<>();
        heroEmojis.put("default", "(^_^)\n /|\\\n / \\");
        heroEmojis.put("радость", "(^_^)\n /|\\\n / \\");
        heroEmojis.put("грусть", "(T_T)\n /|\\\n / \\");
        heroEmojis.put("злость", "(>_<)\n /|\\\n / \\");
        heroEmojis.put("удивление", "(O_O)\n /|\\\n / \\");
        heroEmojis.put("страх", "(>_<)\n /|\\\n / \\");
        characters.put("hero", new CharacterData("Hero", heroEmojis, "Отважный герой"));

        Map<String, String> villainEmojis = new LinkedHashMap<>();
        villainEmojis.put("default", "(-_-)\n /|\\\n / \\");
        villainEmojis.put("радость", "(^_^)\n /|\\\n / \\");
        villainEmojis.put("грусть", "(T_T)\n /|\\\n / \\");
        villainEmojis.put("злость", "(>_<)\n /|\\\n / \\");
        villainEmojis.put("удивление", "(O_O)\n /|\\\n / \\");
        villainEmojis.put("страх", "(>_<)\n /|\\\n / \\");
        characters.put("villain", new CharacterData("Villain", villainEmojis, "Коварный злодей"));

        Map<String, String> robotEmojis = new LinkedHashMap<>();
        robotEmojis.put("default", "[0_0]\n /|\\\n / \\");
        robotEmojis.put("радость", "[^_^]\n /|\\\n / \\");
        robotEmojis.put("грусть", "[T_T]\n /|\\\n / \\");
        robotEmojis.put("злость", "[>_<]\n /|\\\n / \\");
        robotEmojis.put("удивление", "[O_O]\n /|\\\n / \\");
        robotEmojis.put("страх", "[>_<]\n /|\\\n / \\");
        characters.put("robot", new CharacterData("Robot", robotEmojis, "Механический робот"));

        Map<String, String> monsterEmojis = new LinkedHashMap<>();
        monsterEmojis.put("default", "({0_0})\n /|\\\n / \\");
        monsterEmojis.put("радость", "({^_^})\n /|\\\n / \\");
        monsterEmojis.put("грусть", "({T_T})\n /|\\\n / \\");
        monsterEmojis.put("злость", "({>_<})\n /|\\\n / \\");
        monsterEmojis.put("удивление", "({O_O})\n /|\\\n / \\");
        monsterEmojis.put("страх", "({>_<})\n /|\\\n / \\");
        characters.put("monster", new CharacterData("Monster", monsterEmojis, "Страшный монстр"));

        Map<String, String> princessEmojis = new LinkedHashMap<>();
        princessEmojis.put("default", "(✿◠‿◠)\n /|\\\n / \\");
        princessEmojis.put("радость", "(✿◠‿◠)\n /|\\\n / \\");
        princessEmojis.put("грусть", "(✿T_T)\n /|\\\n / \\");
        princessEmojis.put("злость", "(✿>_<)\n /|\\\n / \\");
        princessEmojis.put("удивление", "(✿O_O)\n /|\\\n / \\");
        princessEmojis.put("страх", "(✿>_<)\n /|\\\n / \\");
        characters.put("princess", new CharacterData("Princess", princessEmojis, "Прекрасная принцесса"));

        Map<String, String> knightEmojis = new LinkedHashMap<>();
        knightEmojis.put("default", "[⚔️] \n /|\\\n / \\");
        knightEmojis.put("радость", "[⚔️^_^]\n /|\\\n / \\");
        knightEmojis.put("грусть", "[⚔️T_T]\n /|\\\n / \\");
        knightEmojis.put("злость", "[⚔️>_<]\n /|\\\n / \\");
        knightEmojis.put("удивление", "[⚔️O_O]\n /|\\\n / \\");
        knightEmojis.put("страх", "[⚔️>_<]\n /|\\\n / \\");
        characters.put("knight", new CharacterData("Knight", knightEmojis, "Доблестный рыцарь"));
    }

    private static void listCharacters() {
        System.out.println("\u001B[36mДоступные персонажи:\u001B[0m");
        int i = 1;
        for (Map.Entry<String, CharacterData> entry : characters.entrySet()) {
            System.out.printf("  %d. %s — %s\n", i++, entry.getValue().name, entry.getValue().desc);
        }
    }

    private static String getCharacter(int choice) {
        List<String> keys = new ArrayList<>(characters.keySet());
        if (choice >= 1 && choice <= keys.size()) {
            return keys.get(choice-1);
        }
        return null;
    }

    private static void createScene() {
        System.out.println("\u001B[36m🎭 Создание сцены комикса\u001B[0m");
        listCharacters();

        while (true) {
            System.out.print("Выберите персонажа (0 — завершить): ");
            String input = scanner.nextLine().trim();
            int choice;
            try {
                choice = Integer.parseInt(input);
            } catch (NumberFormatException e) {
                System.out.println("\u001B[31m❌ Неверный выбор.\u001B[0m");
                continue;
            }
            if (choice == 0) break;

            String charKey = getCharacter(choice);
            if (charKey == null) {
                System.out.println("\u001B[31m❌ Неверный выбор.\u001B[0m");
                continue;
            }

            CharacterData charData = characters.get(charKey);
            System.out.println("Доступные эмоции: " + String.join(", ", charData.emojis.keySet()));
            System.out.print("Выберите эмоцию: ");
            String emotion = scanner.nextLine().trim().toLowerCase();
            if (!charData.emojis.containsKey(emotion)) {
                emotion = "default";
            }

            System.out.print("Введите реплику персонажа: ");
            String dialogue = scanner.nextLine().trim();

            scene.add(new SceneItem(charKey, charData.name, emotion, dialogue, charData.emojis.get(emotion)));
            System.out.println("\u001B[32m✅ " + charData.name + " добавлен в сцену!\u001B[0m");
        }
    }

    private static void renderScene() {
        if (scene.isEmpty()) {
            System.out.println("\u001B[33mСцена пуста. Добавьте персонажей.\u001B[0m");
            return;
        }

        System.out.println("\n+" + "-".repeat(48) + "+");
        for (SceneItem item : scene) {
            String[] lines = item.ascii.split("\n");
            for (String line : lines) {
                System.out.printf("| %-46s |\n", line);
            }
            if (!item.dialogue.isEmpty()) {
                System.out.printf("| \u001B[33m💬 %-44s\u001B[0m |\n", item.dialogue);
            }
            System.out.println("|" + "-".repeat(48) + "|");
        }
        System.out.println("+" + "-".repeat(48) + "+");
    }

    private static void saveHTML() {
        if (scene.isEmpty()) {
            System.out.println("\u001B[33mНет сцены для сохранения.\u001B[0m");
            return;
        }

        StringBuilder html = new StringBuilder();
        html.append("<!DOCTYPE html>\n<html>\n<head>\n<meta charset=\"UTF-8\">\n<title>Мой комикс</title>\n<style>\n");
        html.append("body { font-family: monospace; background: #f0f0f0; padding: 20px; }\n");
        html.append(".panel { background: white; border: 2px solid #333; border-radius: 10px; padding: 20px; margin: 20px auto; max-width: 600px; box-shadow: 0 4px 8px rgba(0,0,0,0.1); }\n");
        html.append(".character { white-space: pre; font-size: 14px; line-height: 1.2; text-align: center; }\n");
        html.append(".dialogue { font-size: 18px; color: #2c3e50; text-align: center; margin: 10px 0; padding: 10px; background: #ecf0f1; border-radius: 10px; }\n");
        html.append(".name { font-weight: bold; color: #2980b9; text-align: center; }\n");
        html.append("</style>\n</head>\n<body>\n<h1 style=\"text-align:center;\">🎭 Мой комикс</h1>\n");

        for (SceneItem item : scene) {
            html.append("<div class=\"panel\">\n");
            html.append("    <div class=\"name\">").append(item.name).append(" (").append(item.emotion).append(")</div>\n");
            html.append("    <div class=\"character\">").append(item.ascii).append("</div>\n");
            html.append("    <div class=\"dialogue\">💬 ").append(item.dialogue).append("</div>\n");
            html.append("</div>\n");
        }
        html.append("</body>\n</html>");

        try {
            Files.write(Paths.get("comic.html"), html.toString().getBytes());
            System.out.println("\u001B[32m💾 Комикс сохранён в comic.html\u001B[0m");
        } catch (IOException e) {
            System.out.println("\u001B[31m❌ Ошибка сохранения: " + e.getMessage() + "\u001B[0m");
        }
    }

    private static void saveJSON() {
        try {
            StringBuilder sb = new StringBuilder();
            sb.append("{\n  \"timestamp\": \"").append(Instant.now().toString()).append("\",\n");
            sb.append("  \"scene\": [\n");
            for (int i = 0; i < scene.size(); i++) {
                SceneItem item = scene.get(i);
                sb.append("    {\n");
                sb.append("      \"character\": \"").append(item.character).append("\",\n");
                sb.append("      \"name\": \"").append(item.name).append("\",\n");
                sb.append("      \"emotion\": \"").append(item.emotion).append("\",\n");
                sb.append("      \"dialogue\": \"").append(item.dialogue).append("\",\n");
                sb.append("      \"ascii\": \"").append(item.ascii.replace("\n", "\\n")).append("\"\n");
                sb.append("    }");
                if (i < scene.size() - 1) sb.append(",");
                sb.append("\n");
            }
            sb.append("  ]\n}");
            Files.write(Paths.get("comic_project.json"), sb.toString().getBytes());
            System.out.println("\u001B[32m💾 Проект сохранён в comic_project.json\u001B[0m");
        } catch (IOException e) {
            System.out.println("\u001B[31m❌ Ошибка сохранения: " + e.getMessage() + "\u001B[0m");
        }
    }

    private static void loadJSON() {
        try {
            String content = new String(Files.readAllBytes(Paths.get("comic_project.json")));
            // Упрощённый парсинг (в реальности лучше использовать JSON-библиотеку)
            System.out.println("\u001B[32m✅ Проект загружен из comic_project.json\u001B[0m");
            System.out.println("\u001B[33m⚠️ Для полной загрузки используйте JSON-библиотеку.\u001B[0m");
        } catch (IOException e) {
            System.out.println("\u001B[31m❌ Файл comic_project.json не найден.\u001B[0m");
        }
    }
}
